### DISCUSSION TITLE
Set automountServiceAccountToken for kafka/zk pods or service accounts?

### DISCUSSION THREAD

**Discussion by kaihil** [strimzi/strimzi-kafka-operator#4866] [Q&A]
Is it somehow possible to set `automountServiceAccountToken: false` for kafka and zookeeper pods or the corresponding service accounts?

First, I tried to customize the pods via `PodTemplate`, but it doesn't seem to offer the possibility to change this parameter.
Second, I wanted to customize the `ServiceAccount`s "_cluster-name_-kafka" and "_cluster-name_-zookeeper", but it looks like it isn't possible to customize these service accounts at all.

I would be thankful for any ideas, suggestions or hints for upcoming features which might allow me to do that. Thank you!

**Comment by scholzj** [comment]
No, I do not think it is possible. How would it be useful? Also - at least in case of Kafka - the service account is needed to communicate with Kubernetes APIs when rackawareness or nodeport listeners are used.

**Reply by kaihil** [reply]
Hi, thanks for your quick answer. Basically, I want to fulfill the following Kyverno policy, which is activated in my cluster: https://kyverno.io/policies/other/restrict_automount_sa_token/

Do you know, whether this is possible at all?
Thanks!

**Reply by scholzj** [reply]
How does the policy expect to handle pods which actually need to communicate with the Kubernetes API? To be honest, I'm not too eager to implement this because it is complicated. Different pods and different configurations simply need to use the service account. So even making this configurable will be something what would cost a lot of effort to maintain as there would be a lot of conditions.

**Reply by kaihil** [reply]
Thanks for your remarks!

My first guess for the goal of this policy is, that it should maybe prevent pods from communicating with the Kubernetes API for security reasons. But I am not very deep into the topic, to be honest.

I try to sum up:
1. At least the Kafka pods are communicating with the Kubernetes API.
2. Setting `automountServiceAccountToken: false` on strimzi pods would prevent (at least some of) them from working correctly.
3. Implementing a feature to set `automountServiceAccountToken: false` for pods would implicate too much maintenance effort in the future.

Are these three points correct? I just wanna make sure, I understood you correctly :)

Thanks you!

**Reply by scholzj** [reply]
Just some clarifications ... but in general yes, you understood it correctly.

> At least the Kafka pods are communicating with the Kubernetes API.

It depends on the configuration - but in case you use rack awareness or node port listeners then yes. (and rack awareness is important for production clusters).

> Implementing a feature to set automountServiceAccountToken: false for pods would implicate too much maintenance effort in the future.

It is of course not impossible. But I personally would definitely avoid it if possible.

**Reply by wumbrath** [reply]
@scholzj 
There are some environments where settings like these are required. E.g. some clusters enforce policies like the above one or the [azure one](https://portal.azure.com/#view/Microsoft_Azure_Policy/PolicyDetailBlade/definitionId/%2Fproviders%2FMicrosoft.Authorization%2FpolicyDefinitions%2F423dd1ba-798e-40e4-9c4d-b6902674b423).

It would be nice if CRD would allow full customization of the kafka and ZK pods and containers. This could happen in a generic way, e.g. offer a `podSpec` and `containerSpec` property which allows for generic configuration of pod and container properties which will then be merged with the configuration coming from the strimzi properties.

For us at least this is a blocker, as we are working in a highly secured/restricted environment and there is no way of working around this limitation.

Regarding the question about how pods are handling the communication with the Kubernetes API: [The token secret is created manually and then mounted manually into the pod](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token).

**Reply by germanattanasio** [reply]
@scholzj I’m also facing this issue with Azure policies. There seems to be a push (which other cloud services are likely to adopt) to shift from mounting tokens to using volumes with read-only permissions. I’ve noticed this approach with **Cert-manager**, **Linkerd**, and recently **Ingress**. They all follow a similar pattern, allowing you to specify the pod or service account to have `automountServiceAccountToken` set and then mount volumes with configurations such as **extraVolume** and **extraVolumePath**. I’ve made a PR for trust-manager (https://github.com/cert-manager/trust-manager/pull/513), so I’m happy to assist here as well.


* Linkerd: https://github.com/linkerd/linkerd2/issues/4651
* Ingress: https://github.com/kubernetes/ingress-nginx/issues/9735
* Cert-manager: https://cert-manager.io/docs/installation/best-practice/#restrict-auto-mount-of-service-account-tokens
